streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/DefaultTaskInfo.java

```diff
@@ -0,0 +1,120 @@
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.assignment.TaskInfo;
+
+public class DefaultTaskInfo implements TaskInfo {
+
+    private final TaskId id;
+    private final boolean isStateful;
+    private final Map<TopicPartition, Set<String>> partitionToRackIds;
+    private final Set<String> stateStoreNames;
+    private final Set<TopicPartition> inputTopicPartitions;
+    private final Set<TopicPartition> changelogTopicPartitions;
+
+    public DefaultTaskInfo(final TaskId id,
+                           final boolean isStateful,
+                           final Map<TopicPartition, Set<String>> partitionToRackIds,
+                           final Set<String> stateStoreNames,
+                           final Set<TopicPartition> inputTopicPartitions,
+                           final Set<TopicPartition> changelogTopicPartitions) {
+        this.id = id;
+        this.partitionToRackIds = unmodifiableMap(partitionToRackIds);
+        this.isStateful = isStateful;
+        this.stateStoreNames = unmodifiableSet(stateStoreNames);
+        this.inputTopicPartitions = unmodifiableSet(inputTopicPartitions);
+        this.changelogTopicPartitions = unmodifiableSet(changelogTopicPartitions);
+    }
+
+    public static DefaultTaskInfo of(final TaskId taskId,
+                                     final boolean isStateful,
+                                     final ValueMapper<TopicPartition, String> previousOwnerForPartition,
+                                     final Map<String, Optional<String>> rackForConsumer,
+                                     final Map<TaskId, Set<TopicPartition>> inputPartitionsForTask,
+                                     final Map<TaskId, Set<TopicPartition>> changelogPartitionsForTask) {
+
+        final Set<TopicPartition> inputPartitions = inputPartitionsForTask.get(taskId);
+        final Set<TopicPartition> changelogPartitions = changelogPartitionsForTask.get(taskId);
+        final Map<TopicPartition, Set<String>> racksForPartition = new HashMap<>();
+
+        inputPartitions.forEach(partition -> {
+            racksForPartition.computeIfAbsent(partition, k -> new HashSet<>());
+            final String consumer = previousOwnerForPartition.apply(partition);
+            final Optional<String> rack = rackForConsumer.get(consumer);
+            rack.ifPresent(s -> racksForPartition.get(partition).add(s));
+        });
+
+        changelogPartitions.forEach(partition -> {
+            racksForPartition.computeIfAbsent(partition, k -> new HashSet<>());
+            final String consumer = previousOwnerForPartition.apply(partition);
+            final Optional<String> rack = rackForConsumer.get(consumer);
+            rack.ifPresent(s -> racksForPartition.get(partition).add(s));
+        });
+
+        final Set<String> stateStoreNames = new HashSet<>();
+        return new DefaultTaskInfo(
+            taskId,
+            isStateful, // All standby tasks are stateful.
```

Comment:
Ah, another thing to note here is that this class corresponds to a "logical task", not a "physical one". I just made up those terms but hopefully this will make sense: a "physical" task can be active or standby and represents an actual task that was assigned to a client and will be running on that client, where a "logical" task is just the metadata corresponding to that task id. Where a "task id" logically represents a combination of subtopology (grouping of processors) and partition number. So a logical task doesn't have a concept of active vs standby because it's just metadata, this class is basically telling the assignor which tasks exist in this application. The assignor then has to create a set of physical tasks to actually be assigned, basically one active task and however many standby tasks for each "logical task"

Hope that didn't make things more confusing...anyways this comment isn't incorrect, but it doesn't exactly apply in this context. The "isStateful" is just metadata related to whether it has state stores in this subtopology (I'll tell you how to get this info later, or you can even compute it based on `stateStoresNames#isEmpty`)